Source/items.cpp

```diff
@@ -468,45 +468,49 @@ void SpawnNote()
 
 void CalcSelfItems(Player &player)
 {
+	bool changeflag;
+	// first iteration is used for collecting stat bonuses from items
+	bool firstCheck = true;
 	int sa = 0;
 	int ma = 0;
 	int da = 0;
-	for (int i = 0; i < NUM_INVLOC; i++) {
-		auto &equipment = player.InvBody[i];
-		if (!equipment.isEmpty()) {
-			equipment._iStatFlag = true;
-			if (equipment._iIdentified) {
-				sa += equipment._iPLStr;
-				ma += equipment._iPLMag;
-				da += equipment._iPLDex;
-			}
-		}
-	}
-
-	bool changeflag;
 	do {
 		changeflag = false;
-		auto *pi = player.InvBody;
-		for (int i = 0; i < NUM_INVLOC; i++, pi++) {
-			if (!pi->isEmpty() && pi->_iStatFlag) {
-				bool sf = true;
-				if (sa + player._pBaseStr < pi->_iMinStr)
-					sf = false;
-				if (ma + player._pBaseMag < pi->_iMinMag)
-					sf = false;
-				if (da + player._pBaseDex < pi->_iMinDex)
-					sf = false;
-				if (!sf) {
-					changeflag = true;
-					pi->_iStatFlag = false;
-					if (pi->_iIdentified) {
-						sa -= pi->_iPLStr;
-						ma -= pi->_iPLMag;
-						da -= pi->_iPLDex;
-					}
+
+		// cap stats to 0
+		int currstr = std::max(0, sa + player._pBaseStr);
+		int currmag = std::max(0, ma + player._pBaseMag);
+		int currdex = std::max(0, da + player._pBaseDex);
+
+		for (auto &equipment : player.InvBody) {
+			if (equipment.isEmpty())
+				continue;
+
+			if (firstCheck) {
+				equipment._iStatFlag = true;
+				if (equipment._iIdentified) {
+					sa += equipment._iPLStr;
+					ma += equipment._iPLMag;
+					da += equipment._iPLDex;
 				}
+				continue;
+			}

```

Comment:
This doesn't set `changed` to `true` so the process will terminate after the first iteration and leave all items usable.

Considering @glebm`s suggestion there isn't really anything saved from combining the code at this point, since it would be shorter to have this bit outside again:

```cpp
for (Item &equipment : EquippedPlayerItemsRange(player)) {
	equipment._iStatFlag = true;
	if (equipment._iIdentified) {
		sa += equipment._iPLStr;
		ma += equipment._iPLMag;
		da += equipment._iPLDex;
	}
}
```

Reply:
Other then that it was pretty clever :)

Reply:
How not?
`firstCheck = false;` after the loop

Reply:
The problem is that `changeflag` will be false after the first run so the while loop will just exit before doing the actual checks.

Reply:
good point : D